## Exploratory Data Analysis

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv('heart.csv')

In [3]:
df.rename(columns = {
    'age' : 'Age',
    'sex' : 'Sex',
    'cp' : 'ChestPainType',
    'trestbps' : 'RestingBP',
    'chol' : 'Cholesterol',
    'fbs' : 'Fasting Blood Sugar',
    'restecg' : 'Resting ECG',
    'thalach' : 'Max Heart Rate',
    'exang' : 'Exercise Induced Angina',
    'oldpeak' : 'ST Depression',
    'slope' : 'Slope Of ST Segment',
    'ca' : 'Major Vessels Count',
    'thal' : 'Thalassemia',
    'target' : 'Heart Disease'
}, inplace = True)
df.columns.to_list()

['Age',
 'Sex',
 'ChestPainType',
 'RestingBP',
 'Cholesterol',
 'Fasting Blood Sugar',
 'Resting ECG',
 'Max Heart Rate',
 'Exercise Induced Angina',
 'ST Depression',
 'Slope Of ST Segment',
 'Major Vessels Count',
 'Thalassemia',
 'Heart Disease']

In [4]:
gender_mapping = {0: 'Female', 1: 'Male'}
df['Sex'] = df['Sex'].map(gender_mapping)

cp_mapping = {0: "Typical Angina", 1: "Atypical Angina", 2: "Non-Anginal Pain", 3: "Asymptomatic"}
df['ChestPainType'] = df['ChestPainType'].map(cp_mapping)

heart_mapping = {1: 'Has Heart Disease', 0: 'No Heart Disease'}
df['Heart Disease'] = df['Heart Disease'].map(heart_mapping)

fbs_mapping = {1: 'High (>120mg/dl)', 0: 'Normal'}
df['Fasting Blood Sugar'] = df['Fasting Blood Sugar'].map(fbs_mapping)

exang_mapping = {1: 'Yes', 0: 'No'}
df['Exercise Induced Angina'] = df['Exercise Induced Angina'].map(exang_mapping)

print(df.head(5))

   Age     Sex     ChestPainType  RestingBP  Cholesterol Fasting Blood Sugar  \
0   63    Male      Asymptomatic        145          233    High (>120mg/dl)   
1   37    Male  Non-Anginal Pain        130          250              Normal   
2   41  Female   Atypical Angina        130          204              Normal   
3   56    Male   Atypical Angina        120          236              Normal   
4   57  Female    Typical Angina        120          354              Normal   

   Resting ECG  Max Heart Rate Exercise Induced Angina  ST Depression  \
0            0             150                      No            2.3   
1            1             187                      No            3.5   
2            0             172                      No            1.4   
3            1             178                      No            0.8   
4            1             163                     Yes            0.6   

   Slope Of ST Segment  Major Vessels Count  Thalassemia      Heart Disease  
0 

Renaming columns

In [5]:
df.isnull().sum()

Age                        0
Sex                        0
ChestPainType              0
RestingBP                  0
Cholesterol                0
Fasting Blood Sugar        0
Resting ECG                0
Max Heart Rate             0
Exercise Induced Angina    0
ST Depression              0
Slope Of ST Segment        0
Major Vessels Count        0
Thalassemia                0
Heart Disease              0
dtype: int64

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      303 non-null    int64  
 1   Sex                      303 non-null    object 
 2   ChestPainType            303 non-null    object 
 3   RestingBP                303 non-null    int64  
 4   Cholesterol              303 non-null    int64  
 5   Fasting Blood Sugar      303 non-null    object 
 6   Resting ECG              303 non-null    int64  
 7   Max Heart Rate           303 non-null    int64  
 8   Exercise Induced Angina  303 non-null    object 
 9   ST Depression            303 non-null    float64
 10  Slope Of ST Segment      303 non-null    int64  
 11  Major Vessels Count      303 non-null    int64  
 12  Thalassemia              303 non-null    int64  
 13  Heart Disease            303 non-null    object 
dtypes: float64(1), int64(8), o

In [7]:
df.describe()

,Age,RestingBP,Cholesterol,Resting ECG,Max Heart Rate,ST Depression,Slope Of ST Segment,Major Vessels Count,Thalassemia
count,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000
mean,54.366337,131.623762,246.264026,0.528053,149.646865,1.039604,1.399340,0.729373,2.313531
std,9.082101,17.538143,51.830751,0.525860,22.905161,1.161075,0.616226,1.022606,0.612277
min,29.000000,94.000000,126.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000
25%,47.500000,120.000000,211.000000,0.000000,133.500000,0.000000,1.000000,0.000000,2.000000
50%,55.000000,130.000000,240.000000,1.000000,153.000000,0.800000,1.000000,0.000000,2.000000
75%,61.000000,140.000000,274.500000,1.000000,166.000000,1.600000,2.000000,1.000000,3.000000
max,77.000000,200.000000,564.000000,2.000000,202.000000,6.200000,2.000000,4.000000,3.000000


## 2. Statistical summary
Key observations:
- Dataset contains 303 patients aged 29 - 77 (mean age: 54)
- Average cholestrol is 246 mg/dl (above the healthy threshold of 200)
- 54% of patients in this dataset have heart disease
- No missing values detected — dataset is clean and ready for analysis

## 3. Research Questions
Before diving into analysis, we define the key questions we want to answer:

1. Does age increase the risk of heart disease?
2. Are males or females more at risk?
3. Does high cholesterol contribute to heart disease?
4. Does chest pain type predict heart disease?
5. Does maximum heart rate differ between patients with and without heart disease?
6. Which factors are most strongly correlated with heart disease?

In [8]:
age_analysis = df.groupby('Heart Disease')['Age'].mean()
print("Average age by heart disease status:")
print(f"No Heart Disease (0): {age_analysis['No Heart Disease']:.1f} years")
print(f"Has Heart Disease (1): {age_analysis['Has Heart Disease']:.1f} years")

Average age by heart disease status:
No Heart Disease (0): 56.6 years
Has Heart Disease (1): 52.5 years


In [9]:
gender_analysis = df.groupby('Heart Disease')['Sex'].value_counts()
print("Heart Disease by Gender:")
print(gender_analysis)

Heart Disease by Gender:
Heart Disease      Sex   
Has Heart Disease  Male       93
                   Female     72
No Heart Disease   Male      114
                   Female     24
Name: count, dtype: int64


## 4. Key Findings

### Age Analysis
- Patients with heart disease are slightly younger (52.5 yrs) than those without (56.6 yrs)
- Suggests heart disease affects adults across all age groups, not just the elderly

### Gender Analysis
- 75% of female patients have heart disease vs 45% of male patients
- Women appear at higher risk in this dataset

In [10]:
chol_analysis = df.groupby('Heart Disease')['Cholesterol'].mean()
print("Average Cholesterol by heart disease status:")
print(f"No Heart Disease : {chol_analysis['No Heart Disease']:.1f} mg/dl")
print(f"Has Heart Disease : {chol_analysis['Has Heart Disease']:.1f} mg/dl")

Average Cholesterol by heart disease status:
No Heart Disease : 251.1 mg/dl
Has Heart Disease : 242.2 mg/dl


In [11]:
chest_pain_analysis = df.groupby('ChestPainType')['Heart Disease'].value_counts()
print("Heart Disease by Chest Pain Type:")
print(chest_pain_analysis)

Heart Disease by Chest Pain Type:
ChestPainType     Heart Disease    
Asymptomatic      Has Heart Disease     16
                  No Heart Disease       7
Atypical Angina   Has Heart Disease     41
                  No Heart Disease       9
Non-Anginal Pain  Has Heart Disease     69
                  No Heart Disease      18
Typical Angina    No Heart Disease     104
                  Has Heart Disease     39
Name: count, dtype: int64


In [12]:
max_heart_rate_analysis = df.groupby('Heart Disease')['Max Heart Rate'].mean()
print("Heart Disease by Max Heart Rate:")
print(f"No Heart Disease : {max_heart_rate_analysis['No Heart Disease']:.1f} bpm")
print(f"Has Heart Disease : {max_heart_rate_analysis['Has Heart Disease']:.1f} bpm")

Heart Disease by Max Heart Rate:
No Heart Disease : 139.1 bpm
Has Heart Disease : 158.5 bpm


## 4. Key Findings (continued)

### Cholesterol
- Patients WITHOUT heart disease have slightly higher cholesterol (251 vs 242 mg/dl)
- Cholesterol alone is not a strong predictor of heart disease in this dataset
- Other factors appear more significant

### Chest Pain Type
- Surprisingly, most Typical Angina patients do NOT have heart disease
- Asymptomatic patients are more likely to have heart disease
- Silent heart disease is more dangerous — no symptoms doesn't mean safe!

### Max Heart Rate
- Patients with heart disease achieve higher max heart rates (158 vs 139 bpm)
- Suggests the heart compensates by beating faster in affected patients

In [13]:
df['Heart Disease Numeric'] = df['Heart Disease'].map({'Has Heart Disease': 1, 'No Heart Disease': 0})

correlation = df[['Age', 'Cholesterol', 'Max Heart Rate', 
                   'RestingBP', 'ST Depression',
                   'Heart Disease Numeric']].corr()

print("Correlation with Heart Disease:")
print(correlation['Heart Disease Numeric'].sort_values(ascending=False).round(3))

Correlation with Heart Disease:
Heart Disease Numeric    1.000
Max Heart Rate           0.422
Cholesterol             -0.085
RestingBP               -0.145
Age                     -0.225
ST Depression           -0.431
Name: Heart Disease Numeric, dtype: float64


## 5. Correlation Analysis

Strongest predictors of heart disease:
- ST Depression (-0.431): strongest predictor overall
- Max Heart Rate (+0.422): higher rates associated with heart disease
- Age (-0.225): younger patients more affected
- Cholesterol (-0.085): surprisingly weak predictor

Note: Negative correlation means lower values are associated 
with presence of heart disease

In [14]:
# Outlier Detection using IQR method
Q1 = df[['Age', 'Cholesterol', 'Max Heart Rate', 'RestingBP', 'ST Depression']].quantile(0.25)
Q3 = df[['Age', 'Cholesterol', 'Max Heart Rate', 'RestingBP', 'ST Depression']].quantile(0.75)
IQR = Q3 - Q1

outliers = ((df[['Age', 'Cholesterol', 'Max Heart Rate', 'RestingBP', 'ST Depression']] < (Q1 - 1.5 * IQR)) | 
            (df[['Age', 'Cholesterol', 'Max Heart Rate', 'RestingBP', 'ST Depression']] > (Q3 + 1.5 * IQR))).sum()

print("Number of Outliers per Column:")
print(outliers)

Number of Outliers per Column:
Age               0
Cholesterol       5
Max Heart Rate    1
RestingBP         9
ST Depression     5
dtype: int64


## 6. Outlier Detection

Using the IQR (Interquartile Range) method to detect outliers:

| Column | Outliers |
|---|---|
| Age | 0 |
| Cholesterol | 5 |
| Resting Blood Pressure | 9 |
| ST Depression | 5 |
| Max Heart Rate | 1 |

- Outliers are minimal (less than 3% of dataset)
- Retained in the dataset as they likely represent genuine 
  extreme medical cases rather than data entry errors
- Dataset is clean and reliable for analysis